File: src/navigator.rb
require 'pp'
require 'colorize'
require_relative 'keyboard'

class Navigator
  attr_accessor :map, :position, :x_adj, :y_adj

  def initialize
    @position = Point.new({:x => 0, :y => 0})
    @map = [[Point::CURRENT]]
    @x_adj = 0
    @y_adj = 0
  end

  def exec
    while true
      system('clear')
      show_map
      pp @position
      cmd = read_keyboard
      move(cmd)
    end
  end

  private

  def move(cmd)
    case cmd
    when :up
      destination = Point.new({:x => @position.x - 1, :y => @position.y})
      expand_map(cmd, destination)

      @map[@position.x][@position.y] = Point::VISITED
      @map[destination.x][destination.y] = Point::CURRENT

      @position = destination
    when :right
      destination = Point.new({:x => @position.x, :y => @position.y + 1})
      expand_map(cmd, destination)

      @map[@position.x][@position.y] = Point::VISITED
      @map[destination.x][destination.y] = Point::CURRENT

      @position = destination
    when :down
      destination = Point.new({:x => @position.x + 1, :y => @position.y})
      expand_map(cmd, destination)

      @map[@position.x][@position.y] = Point::VISITED
      @map[destination.x][destination.y] = Point::CURRENT

      @position = destination
    when :left
      destination = Point.new({:x => @position.x, :y => @position.y - 1})
      expand_map(cmd, destination)

      @map[@position.x][@position.y] = Point::VISITED
      @map[destination.x][destination.y] = Point::CURRENT

      @position = destination
    end
  end

  def expand_map(cmd, destination)
    case cmd
    when :up
      return
    when :right
      return if @map.first.size > destination.y + @y_adj
      @map.each do |line|
        line << Point::UNKNOWN
      end
    when :down
      return if @map.size > destination.x + @x_adj
      @map << [Point::UNKNOWN] * @map.first.size
    when :left
      return
    end
  end

  def show_map
    print '   '
    (0...@map.first.size).each do |y|
      print '%3d' % [y + @y_adj]
    end
    puts
    @map.each_with_index do |line, x|
      print '%3d ' % [x + @x_adj]
      line.each_with_index do |pos, y|
        # TODO: add different character for 0,0
        print " o ".on_blue.white if pos == Point::UNKNOWN
        print " o ".on_green.white if pos == Point::VISITED
        print " o ".on_red.white if pos == Point::CURRENT
      end
      puts
    end
  end

  def read_keyboard
    key = Keyboard.read_char

    case key
    when "\e[A"
      return :up
    when "\e[B"
      return :down
    when "\e[C"
      return :right
    when "\e[D"
      return :left
    when "\r"
      return :enter
    when "\u0003"
      exit 0
    end
  end
end

class Point
  UNKNOWN = nil
  CURRENT = 0
  VISITED = 1
  BARRIER = -1

  attr_accessor :x, :y

  def initialize(args)
    @x = args[:x]
    @y = args[:y]
  end
end

navigator = Navigator.new
navigator.exec

File: src/python/navigator.py
from keyboard import Keyboard
from point import Point

class Navigator:
    def __init__(self):
        self.position = Point(0,0)
        self.map = [[Point.CURRENT]]
        self.x_offset = 0
        self.y_offset = 0

    def move(self, cmd):
        if cmd == Keyboard.UP:
            self.__execute_movement(cmd, x_modifier = -1)
        elif cmd == Keyboard.RIGHT:
            self.__execute_movement(cmd, y_modifier = 1)
        elif cmd == Keyboard.DOWN:
            self.__execute_movement(cmd, x_modifier = 1)
        elif cmd == Keyboard.LEFT:
            self.__execute_movement(cmd, y_modifier = -1)

    def show_map(self):
        print "Current Position => x: %d | y: %d" % (self.position.x, self.position.y)
        print "   x "
        print " y   ",
        for y in range(len(self.map[0])):
            print "%3d" % (y + self.y_offset),
        print

        x_count = 0
        for line in self.map:
            print  " %3d  " % (x_count + self.x_offset),
            y_count = 0
            for pos in line:
                if ((x_count + self.x_offset) == 0 and (y_count + self.y_offset) == 0) and pos != Point.CURRENT:
                    print " X ",
                elif pos == Point.UNKNOWN:
                    print " o ",
                elif pos == Point.VISITED:
                    print " o ",
                else:
                    print "<^>",
                y_count += 1
            print
            x_count += 1

    def __execute_movement(self, cmd, x_modifier = 0, y_modifier = 0):
        destination = Point(self.position.x + x_modifier, self.position.y + y_modifier)
        self.__expand_map(cmd, destination)

        self.map[self.position.x - self.x_offset][self.position.y - self.y_offset] = Point.VISITED
        self.map[destination.x - self.x_offset][destination.y - self.y_offset] = Point.CURRENT

        self.position = destination

    def __expand_map(self, cmd, destination):
        if cmd == Keyboard.UP:
            if destination.x > 0: return
            if destination.x >= self.x_offset: return

            self.map.insert(0, [Point.UNKNOWN] * len(self.map[0]))
            self.x_offset -= 1
        elif cmd == Keyboard.RIGHT:
            if destination.y < 0: return
            if destination.y < len(self.map[0]) + self.y_offset: return

            for line in self.map:
                line.append(Point.UNKNOWN)
        elif cmd == Keyboard.DOWN:
            if destination.x < 0: return
            if destination.x < len(self.map) + self.x_offset: return

            self.map.append([Point.UNKNOWN] * len(self.map[0]))
        elif cmd == Keyboard.LEFT:
            if destination.y > 0: return
            if destination.y >= self.y_offset: return

            for line in self.map:
                line.insert(0, Point.UNKNOWN)
            self.y_offset -= 1

